tests/conftest.py

```diff
@@ -20,3 +26,106 @@ def _random_wrapper_applier(arrays, wrapper):
+    if module_parts[0] == "tests" and len(module_parts) > 1:
+        test_module = "/".join(module_parts[1:])
+    else:
+        test_module = request.module.__name__.replace(".", os.sep)
+    test_name = request.node.name
+
+    # Construct path
+    test_dir = base_regr_dir / test_module / test_name
+
+    # Ensure it exists
+    test_dir.mkdir(parents=True, exist_ok=True)
+
+    return test_dir
+
+
+@pytest.fixture
+def original_datadir(lazy_datadir) -> Path:
+    return lazy_datadir
+
+
+@pytest.fixture
+def preserve_obtained(request):
+    """
+    Fixture to preserve obtained files after the test run.
+    """
+    request.node.preserve_obtained = True
+
+
+@pytest.hookimpl(wrapper=True)
+def pytest_runtest_makereport(item, call):
+    """
+    Hook to clean up obtained files. Obtained files are generated
+    in regression tests to compare against the original file.
+    ### This hook ensures that:
+    - If the test passes, and preserve_obtained is not set to True,
+        remove all `*.obtained*` files in the test directory.
+    - If the test fails, do not remove any files. A diff is generated from
+    the html report.
+    """
+    with contextlib.suppress(Exception):
+        if (
+            call.when == "call"  # Only clean up after the test run
+            and call.excinfo is None  # Only clean up if the test passed
+            and not (
+                hasattr(item, "preserve_obtained") and item.preserve_obtained
+            )  # Only clean up if preserve_obtained is not set
+        ):
+            # Test passed clean up obtained files
+            test_module = Path(item.location[0].replace(".py", "")).parts[1:]
+            test_module = os.path.join(*test_module)
+            test_name = item.name
+            test_dir: Path = base_regr_dir / test_module / test_name
+            if test_dir.exists():
+                # find and remove all .obtained files in the directory
+                for file in test_dir.glob("*.obtained*"):
+                    with contextlib.suppress(OSError):
+                        file.unlink()
+
+    return (yield)
+
+
+# Regression fixture for compiler outputs
+@pytest.fixture
+def program_regression(file_regression, tmp_path):
+    """
+    Fixture for program and tree regression testing.
+    This overcomes the challenge of comparing non-deterministic outputs like
+    memory addresses in program trees.
+    """
+
+    def _program_regression(program, formatter=pprint.pformat, extension=".txt"):
+        """
+        Compares the program with the regression fixture.
+        Args:
+            program: The program to compare.
+            formatter: Optional formatter function to convert the AST to a string.
+        """
+        if not isinstance(program, str):
+            program = formatter(program)
+        # Replace memory addresses with ...
+        file_regression.check(
+            re.sub(r"0x[0-9a-fA-F]+", "...", program), extension=extension
```

Comment:
Two comments:

1. I would turn this into an `re.compile` in global scope and then `compiled_re.sub(...)`
2. Can we make this more targeted to specific patters we see? We may not want to replace ALL hex numbers.